### PR TITLE
refactor(print): introduce `UnicodeWidthGraphemes` trait

### PR DESCRIPTION
**Pull request by vladimyr**
#### Description

This introduces a new `Grapheme` type (wrapper for `str`) together with `UnicodeWidthGraphemes` trait that allows calculating grapheme aware widths of strings and graphemes.

#### Motivation and Context

I wanted to abstract away grapheme aware width calculations behind the simple `.width_grapheme()` interface. In order to do that completely, this should probably get moved elsewhere, any suggestions? :thinking: 

#### How Has This Been Tested?

- [ ] I have tested using **MacOS**
- [x] I have tested using **Linux**
- [ ] I have tested using **Windows**

#### Checklist:

- [ ] I have updated the documentation accordingly.
- [ ] I have updated the tests accordingly.

### CODE REVIEW

File: src/print.rs [hunk 1]
```diff
@@ -16,6 +16,44 @@ use crate::module::ALL_MODULES;
...
+
+pub trait UnicodeWidthGraphemes {
+    fn width_graphemes(&self, is_extended: bool) -> usize;
+}
+
+// Assume that graphemes have width of the first character in the grapheme
+impl<'a> UnicodeWidthGraphemes for Grapheme<'a> {
+    #[inline]
+    fn width_graphemes(&self, _is_extended: bool) -> usize {
+        self.0.chars().next().and_then(|c| c.width()).unwrap_or(0)
```
Comment (davidkna) [inline src/print.rs:29]: I think might `max` be slightly more correct:
```suggestion
        self.0.chars().max(|c| c.width()).unwrap_or(0)
```

File: src/print.rs [hunk 2]
```diff
@@ -16,6 +16,43 @@ use crate::module::ALL_MODULES;
...
+    #[inline]
+    fn width_graphemes(&self, is_extended: bool) -> usize {
+        self.graphemes(is_extended)
+            .map(Grapheme)
+            .map(|g| g.width())
+            .sum()
+    }
+}
+
+impl UnicodeWidthGraphemes for String {
```
Comment (davidkna) [inline src/print.rs:41]: If you implement `UnicodeWidthGraphemes` for `AsRef<str>` instead, you should be able to get rid of the extra `impl` here.

**Review by davidkna — COMMENTED** [review]
Overall approach LGTM!

File: src/print.rs (hunk 2 above)
Comment (vladimyr) [reply src/print.rs:41]: I was looking for that! Thank you!

File: src/print.rs [hunk 3]
```diff
@@ -16,6 +16,43 @@ use crate::module::ALL_MODULES;
 use crate::modules;
 use crate::segment::Segment;
 
+pub struct Grapheme<'a>(&'a str);
+
+impl<'a> Grapheme<'a> {
+    pub fn width(&self) -> usize {
+        self.0.chars().flat_map(|c| c.width()).max().unwrap_or(0)
```
[the remark this answers is not in the document]
Comment (vladimyr) [reply src/print.rs:23]: Indeed :+1:

File: src/print.rs [hunk 4]
```diff
@@ -16,6 +16,43 @@ use crate::module::ALL_MODULES;
...
+            .map(Grapheme)
+            .map(|g| g.width())
+            .sum()
+    }
+}
+
+impl UnicodeWidthGraphemes for String {
+    #[inline]
+    fn width_graphemes(&self, is_extended: bool) -> usize {
+        self.as_str().width_graphemes(is_extended)
```
[the remark this answers is not in the document]
Comment (vladimyr) [reply src/print.rs:44]: I wanted to leave it configurable but you are probably right, we only use extended anyway.

**Comment by vladimyr** [comment]
FWIW I'm still feelin' meh about keeping it inside the print module :shrug:

**Review by davidkna — APPROVED** [review]

**Comment by davidkna** [comment]
`utils.rs` maybe? But it doesn't feel much better.

**Comment by vladimyr** [comment]
> `utils.rs` maybe? But it doesn't feel much better.

You've said it all 🤷 Let's leave it as is and move it later on when/if someone comes with a better idea.

**Comment by andytom** [comment]
Thank you as always @vladimyr and thanks for reviewing @davidkna.

**Pull request merged**